### ISSUE TITLE
nixos/acme: force-renewing certificates is unreasonably difficult

### ISSUE THREAD

**Issue by emilazy** [NixOS/nixpkgs#81634]
Right now, the only way I've found that works is to set `validMinDays = 999;` to force the renewal. This wouldn't matter so much (it's usually only necessary if you e.g. tweak certificate options like OCSP Must-Staple) if not for the fact that [Let's Encrypt screwed up](https://arstechnica.com/information-technology/2020/03/lets-encrypt-revoking-https-certs-due-to-certificate-authority-bug/) and now a bunch of people have to do it by tomorrow, which they won't. Oops.

I'm filing this as an issue rather than a PR in part because I'm not really sure what the a good interface would be here; the best thing I can imagine is something like `nix run -f '<nixpkgs/nixos>' something -c force-renew-certs`, which seems weird. Does anyone know of prior precedent for interfaces here?

cc @aanderse @arianvp @m1cr0man @yegortimoshenko

## If you just want to know how to force-renew your certificates this time

Add `security.acme.validMinDays = 999;` to your configuration and run a `nixos-rebuild switch`. This may or may not automatically renew the certificate depending on your nixpkgs version; to make sure, do `systemctl start 'acme-*.service'`. **Make sure to remove the `validMinDays` option and run `nixos-rebuild switch` again afterwards,** or you'll hammer the Let's Encrypt servers for a renewal every day!

**Comment by emilazy** [comment]
(maybe `systemctl start acme-force-renew`? But that feels like a hack.)

**Comment by emilazy** [comment]
Interface/implementation sketch, after discussing with @yegortimoshenko: `acme-force-renew-${domain}.service` omits the `validMinDays` option handling so that lego always renews the certificate. We could then also hook things up so that it automatically runs when options like `ocspMustStaple` are changed.

This would require some reorganization in the acme module to avoid duplicating the service logic, so it would probably be a good idea to clean it up at the same time. I'll try and get around to doing it if nobody else does, but don't want to block anyone who feels like taking it on themselves.

**Comment by emilazy** [comment]
Would also be a good idea to [check for revocation with OCSP](https://raymii.org/s/articles/OpenSSL_Manually_Verify_a_certificate_against_an_OCSP.html) on the timer and do a force-renewal if so; this would mitigate the impact of future mass revocations in that the certificate would only be invalid for a day or so.

**Comment by dhess** [comment]
There's another issue related to this: if you add an `extraDomain` (e.g., you just added a CNAME for the server), the `acme` service doesn't seem to notice the change, and it won't automatically renew as it should.

**Comment by reanimus** [comment]
Changing the ACME server endpoint is also a scenario leading to a forced reload due to acme not noticing the change, too.

**Comment by m1cr0man** [comment]
This issue is effectively solved now. With #91121, you can run `systemctl clean acme-$domain.service` to clear all certificates, and the next start of the service will acquire new ones. It is also documented in the manual too: https://nixos.org/manual/nixos/stable/index.html#module-security-acme-regenerate

**Comment by supermarin** [comment]
Ran into this recently, luckily on a personal website that's being rebuilt right now.
Not sure if I'm doing something wrong, but can't seem to get this working:
```
[root@nix:/home/git]# find / -name supermar.in
/var/www/supermar.in
/var/lib/acme/supermar.in
/var/lib/acme/.lego/supermar.in
^C

[root@nix:/home/git]# ls -la /var/lib/acme/supermar.in/
total 24
drwx------ 2 nginx nginx 4096 Nov 21 19:19 .
drwxr-xr-x 6 root  root  4096 Nov 21 19:25 ..
lrwxrwxrwx 1 nginx nginx   13 Nov 21 19:19 cert.pem -> fullchain.pem
-rw------- 1 nginx nginx 1680 Nov 21 19:19 chain.pem
-rw------- 1 nginx nginx 3262 Nov 21 19:19 fullchain.pem
-rw------- 1 nginx nginx 3489 Nov 21 19:19 full.pem
-rw------- 1 nginx nginx  227 Nov 21 19:19 key.pem

[root@nix:/home/git]# systemctl clean acme-supermar.in.service
Failed to clean unit acme-supermar.in.service: No matching resources found.
```
```
[root@nix:/home/git]# systemctl status acme-supermar.in.service
● acme-supermar.in.service - Renew ACME Certificate for supermar.in
   Loaded: loaded (/nix/store/srj75amm6pbgmxnil6r3yab66vp9gwjj-unit-acme-supermar.in.service/acme-supermar.in.service; enabled; vendor preset:>
   Active: inactive (dead) since Fri 2020-12-04 20:48:20 UTC; 6s ago
  Process: 23860 ExecStart=/nix/store/2n4f9v9bzv95hg3zip258g2nf5ahv5cs-acme-start (code=exited, status=0/SUCCESS)
  Process: 23866 ExecStartPost=/nix/store/h75f7si22692znj500v3gwbbzia08rr8-acme-post-start (code=exited, status=0/SUCCESS)
 Main PID: 23860 (code=exited, status=0/SUCCESS)
       IP: 4.6K in, 978B out
      CPU: 66ms

Dec 04 20:48:20 nix systemd[1]: Starting Renew ACME Certificate for supermar.in...
Dec 04 20:48:20 nix 2n4f9v9bzv95hg3zip258g2nf5ahv5cs-acme-start[23860]: 2020/12/04 20:48:20 [supermar.in] The certificate expires in 76 days, >
Dec 04 20:48:20 nix systemd[1]: acme-supermar.in.service: Succeeded.
Dec 04 20:48:20 nix systemd[1]: Started Renew ACME Certificate for supermar.in.
Dec 04 20:48:20 nix systemd[1]: acme-supermar.in.service: Consumed 66ms CPU time, received 4.5K IP traffic, sent 978B IP traffic.
```

**Comment by pkern** [comment]
Same here. No matching resources when trying to clean.

**Comment by supermarin** [comment]
@pkern just for quickly unblocking, `rm -rf /var/lib/acme*` and `nixos-rebuild` should get you going

**Comment by m1cr0man** [comment]
Hey folks! Sorry for not responding sooner. I read over the generated service config and the systemctl docs, and I believe you need to add the `--what=state` argument like so:

`systemctl clean --what=state acme-m1cr0man.com.service`

I believe this isn't highlighted in the nixos acme docs so I'll fix that in my next PR.
